### DISCUSSION TITLE
Rust Part Two

### DISCUSSION THREAD

**Discussion by Pshaowang** [o3de/o3de#2504] [Developer Articles]
## Topic

This series is a gentle introduction to the concepts of the Rust programming language, and why it might be useful to learn something about it. This part of the series is focused on data structures, method implementations, and error handling.

## Why?

If you’ve read the first part of this series, you already know that Rust is meant to be a full C++ replacement language, with an emphasis on safety and assisting the programmer when they make easily resolved mistakes through useful compiler messages and diagnostics.

(To see the Rust Part One article, please visit https://forums.awsgametech.com/t/new-introduction-to-rust-part-1/10682) 

But you can’t do a lot with Rust just knowing these things. In this part of the series, you’ll be introduced to Rust’s data structure types and methods on data structures, as well as the `Option` and `Result` enumeration types. This will get you started understanding more about Rust’s type system and how to write practical programs, and lead into the topic of the next part in this series.

*Every example in this article was written in, and can be run from, the [Rust Playground](https://play.rust-lang.org/). Try the examples and see what the compiler tells you!*

## Data structures

Rust offers data structures similar to C through the `struct` keyword. Structs are *not* necessarily C-compatible and Rust doesn’t make guarantees about data formatting - the compiler might optimize data layouts. Adding the `#[repr(C)]` directive before a struct guarantees C-compatible structs. Rust even supports field-packing, `union` (with improved support in recent releases!), and there are useful crates (Rust modules) for common C programming paradigms like [bitfields](https://crates.io/crates/modular-bitfield). Here’s an example of a Vector3 type, and a union demonstrating how to split an unsigned 16-bit integer across two 8-bit integers in a union.

```
// 1. Debug - used to print in the {:?} string format placeholder in a later example
// 2. PartialEq - Used for the `assert_eq` comparison
// 3. Copy, Clone - Used to avoid "partially borrowed" errors
#[derive(Debug, PartialEq, Copy, Clone)]
struct Vector3 {
    x: f64,
    y: f64,
    z: f64,
}

// Remove the following line and see what happens!
#[derive(Copy, Clone)]
// Force C-style data alignment to map the bits correctly in the union -
// we don't know what Rust will do on its own, even if by default it
// will "probably" do the right thing for POD structs, especially in a union.
#[repr(C)]
struct RegisterPair {
    f: u8,
    a: u8,
}

union AdjoinedRegisters {
    af: u16,
    pair: RegisterPair,
}
```

How do structs and unions interact with the borrow checker? One question you might have is “What if I only want to borrow *part* of a `struct`?” You can - and sometimes even at the same time as the whole `struct` itself, including overlapping references to the same interior data.

```
fn main() {
    let mut v = Vector3 {
        x: 1.0,
        y: 1.0,
        z: 0.0,
    }; // Struct construction syntax
    let z_ref = &mut v.z;
    *z_ref = 1.0;

    let y_ref = &mut v.y;
    *y_ref = 0.0;

    let v_ref = &mut v;
    v_ref.x = 2.0;
    v_ref.y = 1.5;

    // Unions are INHERENTLY unsafe, because they're two different
    // (not necessarily overlapping) types of data in the same memory.
    // So since the borrow checker is turned off... do what you want,
    // but know what you do, and keep the scope as small as you can!
    unsafe {
        let mut registers = AdjoinedRegisters { af: 0 };
        let af_ref = &mut registers.af;
        *af_ref = 0b0000_0001_1011_0000; // Yes, you can write integer literals as bits!

        let f_ref = &mut registers.pair.f;
        *f_ref = 0;

        print!("{}", registers.af);
    }
}
```

*(If you paste into the [Rust playground](https://play.rust-lang.org/), be sure and include the struct definitions not repeated here.)*

This wasn’t always the case! A major pain point in Rust was once fighting the borrow checker over (perfectly safe) borrows of interior data while borrowing the whole struct in the same scope. It *seems* the language has been updated to only complain when borrows of this type cross lifetimes, or when you violate the mutable/immutable reference rules. Finding an example of a situation where the borrow checker *does* complain about this is a fun(?) exercise left to the reader.

The moral of the story? You don’t need to memorize *every* borrow checker rule - trust it, and be prepared to unlearn cumbersome workarounds as the language becomes more ergonomic. Long-time Rust aficionados even remember an era when you had to manually annotate every reference with a lifetime, no matter where it appeared (and it’s even still valid Rust to do this.)

**Note on subclassing:** Because ‘*subclassing*’ is inherently an OOP paradigm and Rust isn’t interested in forcing OOP on the programmer, Rust doesn’t have the concept of a “*subclass*”. Dynamic typing and “subclass-like behavior” will be covered in Part 3.

## Enumerations, `None`, and error handling

Rust also offers *discriminated enumerations* (aka *tagged unions*) which, if you’ve never encountered them, are great tools for flow control and associated data. Associated data in an `enum` is just a value supplied at the time of creation. You can mix and match between non-discriminated and discriminated enum values, even!

Rather than a contrived example, it’s worth examining the two most important `enum` types in Rust: `Option<T>` and `Result<T, E>`.

```
enum Option<T> {
    Some(T),
    None,
}

enum Result<T, E> {
    OK(T),
    Err(E),
}
```

*(Sidebar: The* `Name<T>` *syntax is, as you’d expect, for generics. Generics and Rust traits will be covered in detail in Part 3.)*

The purpose of `Option<T>` is to replace `NULL` in the language, and this is because `NULL` (in C) can mean whatever we want. Normally we use it as a sentinel (flow control) or soundness check (memory safety), but honestly, nothing’s stopping you. That’s the power of C! This is to say nothing of the kinds of problems *not* checking for null can result in.

Rust’s use of `Option` means it’s never ambiguous when an operation could result in “no valid result”. There’s even convenient syntax for unwrapping the value in an option and handling it directly, with `if let`:

```
fn normalize(vec: &Vector3) -> Option<Vector3> {
    // Please pretend there is a proper floating-point epsilon check here. Thank you.
    if vec.x == 0.0 && vec.y == 0.0 && vec.z == 0.0 {
        return None;
    }

    let normal = f64::sqrt(vec.x.powf(2.0) + vec.y.powf(2.0) + vec.z.powf(2.0));
    Some(Vector3 {
        x: vec.x / normal,
        y: vec.y / normal,
        z: vec.z / normal,
    })
}

fn main() {
    let zero = Vector3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    };
    let vec = Vector3 {
        x: 0.0,
        y: 2.0,
        z: 3.0,
    };

    let result = normalize(&vec);
    // This construction reads as:
    // "If `result` is a `Some` with data, let `normal` be the value of the data."
    // If the result is _not_ a Some, the block is not executed.
    if let Some(normal) = result {
        print!("{:?}", normal);
    }

    let result = normalize(&zero);
    // Because Rust is just so dang good at static analysis it's idiomatic
    // to use the same name for the unwrapped value and the option itself
    // in an `if let`.
    if let Some(result) = result {
        print!("Oops! Invalid result {:?}", result);
    }

    assert_eq!(None, result);
}
```

*(If you paste into the [Rust playground](https://play.rust-lang.org/), be sure and include the struct definitions not repeated here.)*

The `Result<T, E>` type is similar: It’s either an operation that passed with an expected result (`Ok(T)`) or some form of error (`Err(E)`). The convention for any operation which could *fail* (as opposed to *not return a value*) is to return a `Result` of some type.

```
fn normalize(vec: &Vector3) -> Result<Vector3, &'static str> {
    // Please pretend there is a proper floating-point epsilon check here. Thank you.
    if vec.x == 0.0 && vec.y == 0.0 && vec.z == 0.0 {
        return Err("Oh noooooo it's the zero vector");
    }

    let normal = f64::sqrt(vec.x.powf(2.0) + vec.y.powf(2.0) + vec.z.powf(2.0));
    Ok(Vector3 {
        x: vec.x / normal,
        y: vec.y / normal,
        z: vec.z / normal,
    })
}

fn main() {
    let zero = Vector3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    };
    let vec = Vector3 {
        x: 0.0,
        y: 2.0,
        z: 3.0,
    };

    // How about some common uses of the Result type?
    // Result::expect() returns the value wrapped in Ok() and panics if Err
    let result = normalize(&vec).expect("Oh no an unexpected error!");
    println!("{:?}", result);

    // And Result::expect_err() returns the value wrapped in Err() and panics if Ok
    println!(
        "{:?}",
        normalize(&zero).expect_err("Oh no an unexpected OK!")
    );
}
```

*(If you paste into the [Rust playground](https://play.rust-lang.org/), be sure and include the struct definitions not repeated here.)*

This is the first time strings have been used in these articles, because Rust strings are **extremely complicated**. What you need to know here is that string literals are of the* `&str` *type and have* `'static` *lifetime, since they’re compiled into the binary and thus available for the whole lifetime of the program.

Error types do have some restrictions, which will be covered in the next part of the series when *trait bounds* are discussed. But most programmers, especially early in development, happily use `&str` as an error type since all they want is some kind of useful message and it meets all the required criteria.

Of course, people could just willingly ignore `Result` returns! This is what led to the `#[must_use]` attribute: If a value of a must-use type is ever constructed, *it must be used in some manner or the compiler rejects the program.* Naturally, `Result` was declared to be `#[must_use]`. This is Rust’s way of forcing the programmer into either handling errors, or admitting through code that they really don’t care about them or any resulting crashes.

## Functions on user-defined types (methods)

Methods aren’t declared as part of your type: Rather, you *implement methods on a type*. If this sounds more like C-style thinking to you, that’s because it is. Rust isn’t interested in promoting object-oriented programming as a style - which C++ was so deeply tied to - so a more flexible approach to data manipulation is necessary. Let’s add a constructor and a couple of basic operations to `Vector3`.

```
struct Vector3 {
    x: f64,
    y: f64,
    z: f64,
}

impl Vector3 {
    // Rust has no official constructor syntax - structs aren't objects! -
    // but the convention is a static new() implementation, returning an
    // initialized struct. Methods which begin without a `self`
    // argument are static.
    pub fn new(x: f64, y: f64, z: f64) -> Vector3 {
        // If variable names and types match struct fields, the compiler
        // knows what to do!
        Vector3 { x, y, z }
    }

    // Note that for implementations, you want to _borrow_ self. Otherwise,
    // self will follow the normal Rust rules - and move into the scope of
    // the function, and be dropped (deallocated) on function return.
    pub fn dot(&self, v: &Vector3) -> f64 {
        self.x * v.x + self.y * v.y + self.z * v.z
    }

    pub fn normal(&self) -> f64 {
        // Mathematician style:
        f64::sqrt(self.x.powf(2.0) + self.y.powf(2.0) + self.z.powf(2.0))

        // Functional style:
        // (self.x.powf(2.0) + self.y.powf(2.0) + self.z.powf(2.0)).sqrt()
    }

    // If you want to invalidate the original object, you can _consume_ self,
    // but should do so in a way that allows method chaining or returns a
    // reasonable value. This method could _also_ be implemented as
    // normalize(&mut self) - it depends on whether or not you want to
    // prioritize invariance or try and force optimizations you'd get
    // from in-place data manipulation. Rust may make those optimizations anyway,
    // it's pretty good at inlining!
    //
    // However, if you would _reassign_ to self, rewriting the whole object,
    // it's considered good practice to consume and return a new struct.
    //
    // Note the result type `Self` - this is good practice for any function
    // which returns the type being implemented, and even required in some
    // situations.
    pub fn normalize(self) -> Option<Self> {
        let normal = self.normal();
        if normal == 0.0 {
            return None;
        }

        Some(Vector3 {
            x: self.x / normal,
            y: self.y / normal,
            z: self.z / normal,
        })
    }
}
```

The use of `f64::sqrt` vs `(expression as f64).sqrt()` in the example reveal something you might expect - `impl` and `self` are syntax sugar that enable the C++-style method calling in addition to helping the type checker, but you can always fall back on namespaces and passing arguments, similar to how Python method calls work.

Think of `struct`, `union`, or `enum` as “generate the data grouping” and `impl` as “generate a function call table” - just like how in C you’d generally have one data struct and one function table struct. This is a definitely a simplification of what’s happening behind the scenes (especially when traits come in) but is a useful mental model, especially for functional programming.

And yes, you can declare `impl`s on `enum` and `union` types, like you may have guessed from the earlier examples! Check out the [Result documentation](https://doc.rust-lang.org/std/result/enum.Result.html) to see how useful doing this can be.

**Note on access specifiers**: Rust’s `impl` methods and `struct` data are *private by default* and can be made public with the `pub` keyword. The concepts of `friend` and `protected` are enforced by the Rust module system, covered in a future part of the series.

## Next?

With basic data structures and function implementation out of the way, the next article will get into the Rust trait system and generics, which you’ve already seen examples of with `Option<T>` and `Result<T,E>`. Rust has *incredibly extensive* trait and generics support, supported by compile-time checks and a runtime system.

### Official resources on these topics

* Structs - [Using Structs to Structure Related Data - The Rust Programming Languag...](https://doc.rust-lang.org/book/ch05-00-structs.html)
* Enums (including a section on flow control, not covered here) - [Enums and Pattern Matching - The Rust Programming Language](https://doc.rust-lang.org/book/ch06-00-enums.html)
* Methods - [Method Syntax - The Rust Programming Language](https://doc.rust-lang.org/book/ch05-03-method-syntax.html)
* `Option<T>` documentation - [std::option - Rust](https://doc.rust-lang.org/std/option/)
* `Result<T,E>` documentation - [std::result - Rust](https://doc.rust-lang.org/std/result/)

### For the implementation-curious

* Rust’s Foreign Function Interface (FFI) - https://doc.rust-lang.org/nomicon/ffi.html
* Type layouts - https://doc.rust-lang.org/reference/type-layout.html
* Current most active ABI proposal - https://internals.rust-lang.org/t/a-stable-modular-abi-for-rust/12347

Author: Stephen Tramer [stramer]

*Disclaimer: The views expressed here are those of the individual author and do not represent those of Lumberyard or Amazon.*

**Comment by Dimous** [comment]
So, will Rust be integrated?
